### PR TITLE
Use offsetalator in cudf::strings::replace functions

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Adds offsetalator in place of hardcoded offset size_type arrays to the strings replace functions.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/strings/replace/multi.cu [hunk 1]
```diff
@@ -270,92 +290,73 @@ std::unique_ptr<column> replace_character_parallel(strings_column_view const& in
...
+    *d_strings,
+    cudf::detail::offsetalator_factory::make_input_iterator(input.offsets(), input.offset()),
+    d_targets,
+    d_replacements,
+  };
+
+  // Count the number of targets in the entire column.
+  // Note this may over-count in the case where a target spans adjacent strings.
+  auto target_count = thrust::count_if(
+    rmm::exec_policy(stream),
```
Comment (bdice) [inline cpp/src/strings/replace/multi.cu:311]: We can probably use a nosync execution policy in places like this.

File: cpp/src/strings/replace/replace.cu [hunk 2]
```diff
@@ -52,531 +49,388 @@ namespace detail {
...
-  size_type const target_size{};
+  __device__ inline bool is_valid(size_type idx) const { return d_strings.is_valid(idx); }
 
-  __device__ bool operator()(size_type overlap_idx) const
+  /**
+   * @brief Returns true if the target string is found at the given byte position
+   * in the input strings column and is legally within a string row
+   *
+   * @param idx Index of the byte position in the chars column
+   * @param chars_bytes Number of bytes in the chars column
```
Comment (bdice) [inline cpp/src/strings/replace/replace.cu:81]: I only see one parameter. Where is chars_bytes?

File: cpp/src/strings/replace/replace.cu [hunk 3]
```diff
@@ -52,531 +49,388 @@ namespace detail {
...
-    while (first_overlap_idx > 0) {
-      size_type prev_pos_idx = d_overlap_pos_indices[--first_overlap_idx];
-      if (prev_pos_idx + 1 != first_pos_idx) { break; }
-      first_pos_idx = prev_pos_idx;
-    }
+  /**
+   * @brief Returns true if the target string found at the given byte position
+   *
+   * @param idx Index of the byte position in the chars column
+   * @param chars_bytes Number of bytes in the chars column
```
Comment (bdice) [inline cpp/src/strings/replace/replace.cu:104]: Same as above, no chars_bytes parameter?

File: cpp/src/strings/replace/replace.cu [hunk 4]
```diff
@@ -52,531 +49,388 @@ namespace detail {
...
-  if (target_count == 0) {
-    // nothing to replace, copy the input column
-    return std::make_unique<cudf::column>(strings.parent(), stream, mr);
-  }
+struct replace_fn {
+  column_device_view const d_strings;
+  string_view d_target;
+  string_view d_replacement;
+  cudf::size_type maxrepl;
+  int32_t* d_offsets{};
```
Comment (bdice) [inline cpp/src/strings/replace/replace.cu:349]: Is this supposed to be size_type?

File: cpp/src/strings/replace/replace.cu [hunk 5]
```diff
@@ -52,531 +49,388 @@ namespace detail {
...
+          out_ptr = copy_and_increment(out_ptr, in_ptr + lpos, spos - lpos);
+          out_ptr = copy_string(out_ptr, d_repl);
+          lpos    = spos + d_tgt.size_bytes();
+        }
+        spos += d_tgt.size_bytes() - 1;
+        --max_n;
+      }
+      ++spos;
+    }
+    if (out_ptr)  // copy remainder
```
Comment (bdice) [inline cpp/src/strings/replace/replace.cu:385]: I’d prefer to wrap these if/else clauses in braces.

**Review by bdice — COMMENTED** [review]
A few small comments but nothing blocking. Happy to approve once you take a look.

File: cpp/src/strings/replace/replace.cu (hunk 4 above)
Comment (davidwendt) [reply cpp/src/strings/replace/replace.cu:349]: Yep. Pretty old code that just got moved here.

**Review by bdice — APPROVED** [review]
LGTM

**Review by vyasr — APPROVED** [review]

**Comment by vyasr** [comment]
Sorry I missed this earlier. Approved.

**Pull request merged**